### DISCUSSION TITLE
ModbusClient/src/RTUutils.cpp:131:21: error: comparison between distinct pointer types 'HardwareSerial*' and 'HWCDC*' lacks a cast

### DISCUSSION THREAD

**Discussion by BrotherCreamy** [eModbus/eModbus#248] [Q&A]
Hi guys,

I'm getting this build error which occurs during the build of the library:

`ModbusClient/src/RTUutils.cpp:131:21: error: comparison between distinct pointer types 'HardwareSerial*' and 'HWCDC*' lacks a cast`

Any ideas what I'm doing wrong?

Running PlatformIO with Arduino framework.

Cheers,
Brother

**Comment by Miq1** [comment]
The path looks like you have modified the structure of the library - would be helpful to see more of it. Besides that, there is no type called ``HWCDC`` in our code; it again must stem from your application, f.i. when calling ``ModbusclientRTU::begin()``.

**Comment by BrotherCreamy** [comment]
Hi, thanks for your reply.

I have not modified the path, only installed ModbusClient as per the instructions by including the following in my platform.ini:
```
lib_deps = 
  ModbusClient=https://github.com/eModbus/eModbus.git
```

Serial is defined as type HWCDC by the Arduino ESP32 framework (I'm using ESP32) in `/.platformio/packages/framework-arduinoespressif32/cores/esp32/HWCDC.cpp`:
```
#if ARDUINO_USB_MODE
#if ARDUINO_USB_CDC_ON_BOOT//Serial used for USB CDC
HWCDC Serial;
#else
HWCDC USBSerial;
#endif
#endif
```

So I guess there is some confusion there that might only occur with ESP platforms with native USB (like the ESP32-S3 I am using)?

Any recommendation for getting around this?

**Reply by Miq1** [reply]
I see. The bit misleading me was the lib_deps name of ModbusClient you chose - I never tried to change it from eModbus myself 😉 

I am afraid we are not expecting anything but a proper ``HardwareSerial`` here. I never had tried to use the USBserial instead, since I normally am using the "older" ESP32s with three Serial ports. I will have to take a deeper look into this.

**Comment by BrotherCreamy** [comment]
Some more info:

There is also HardwareSerial.cpp, which contains the following:

```
#if !defined(NO_GLOBAL_INSTANCES) && !defined(NO_GLOBAL_SERIAL)
#if ARDUINO_USB_CDC_ON_BOOT //Serial used for USB CDC
HardwareSerial Serial0(0);
#else
HardwareSerial Serial(0);
#endif
#if SOC_UART_NUM > 1
HardwareSerial Serial1(1);
#endif
#if SOC_UART_NUM > 2
HardwareSerial Serial2(2);
#endif
```

which seems to be in contradiction with HWCDC.cpp

```
#if ARDUINO_USB_MODE
#if ARDUINO_USB_CDC_ON_BOOT//Serial used for USB CDC
HWCDC Serial;
#else
HWCDC USBSerial;
#endif
#endif
```

...although maybe I'm misinterpreting it.

Should the serial used for USB CDC be `Serial` or `Serial0`?

And then I suppose that Modbus should use whichever is not used by USB, right?

**Reply by Miq1** [reply]
``Serial`` it should be in my opinion.

**Comment by dsilletti** [comment]
same here, I used the Serial1 (UART1) for the Modbus RTU and changed the Serial (UART0) to use the USB CDC on the ESP32-S3 (for flashing and Serial console purpose only).

To avoid the comparison between  HardwareSerial and HWCDC pointers, I commented the check on the Serial (UART0), leaving in place the check for the Serial1 and Serial2.

I will never use the first Serial (UART0) for the Modbus RTU so for me it's working fine.

```
// UARTinit: modify the UART FIFO copy trigger threshold 
// This is normally set to 112 by default, resulting in short messages not being 
// recognized fast enough for higher Modbus bus speeds
// Our default is 1 - every single byte arriving will have the UART FIFO
// copied to the serial buffer.
int RTUutils::UARTinit(HardwareSerial& serial, int thresholdBytes) {
  int rc = 0;
#if NEED_UART_PATCH
  // Is the threshold value valid? The UART FIFO is 128 bytes only
  if (thresholdBytes > 0 && thresholdBytes < 128) {
    // Yes, it is. Try to identify the Serial/Serial1/Serial2 the user has provided.
    uart_dev_t *uart = nullptr;
    uint8_t uart_num = 99;
    // if (&serial == &Serial) {
    //   uart_num = 0;
    //   uart = &UART0;
    // } else {
      if (&serial == &Serial1) {
        uart_num = 1;
        uart = &UART1;
      } else {
        if (&serial == &Serial2) {
          uart_num = 2;
          uart = &UART2;
        }
      }
    //}
    // Is it a defined serial?
    if (uart_num != 99) {
      // Yes. get the current value and set ours instead
      rc = uart->conf1.rxfifo_full_thrhd;
      uart->conf1.rxfifo_full_thrhd = thresholdBytes;
      LOG_D("Serial%u FIFO threshold set to %d (was %d)\n", uart_num, thresholdBytes, rc);
    } else {
      LOG_W("Unable to identify serial\n");
    }
  } else {
    LOG_E("Threshold must be between 1 and 127! (was %d)", thresholdBytes);
  }
#endif
  // Return the previous value in case someone likes to see it.
  return rc;
}
```
